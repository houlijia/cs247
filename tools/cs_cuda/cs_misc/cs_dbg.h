#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
void dbg_pdata_i( char *s, int *dp, int size ) ;
void dbg_pdata_c( char *s, char *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( char *s, int *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
void dbg_p_d_data_i_mn_skip ( char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_i_mn_v2 ( char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

// tsting ...
void dbg_p_d_data_ll ( char *s, long long *dp, int size ) ;

#endif 
