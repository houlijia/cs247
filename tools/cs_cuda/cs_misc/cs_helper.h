#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void clear_device_mem_i ( int *d_a, int n ) ;
void set_device_mem_i ( int *d_a, int n, int val ) ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;
void h_expand_c_to_i ( char *din, int *dout, int n ) ;	

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_i ( int *dp, int *hp, int size ) ;

#endif 
