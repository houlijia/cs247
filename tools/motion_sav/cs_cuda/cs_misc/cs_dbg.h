#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
