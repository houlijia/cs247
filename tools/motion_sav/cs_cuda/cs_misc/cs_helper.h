#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
