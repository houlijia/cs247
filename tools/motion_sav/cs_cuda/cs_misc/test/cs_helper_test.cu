#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_helper.h"

#define CUDA_DBG
#define BUF_SIZE_INT		10

int *dp1, *dp2, *dp3 ;
float *fp1 ;

int buf1[ BUF_SIZE_INT ] ; 
float dbuf[ BUF_SIZE_INT ] ;

char cbuf1[ BUF_SIZE_INT ] ;

/*
   exptected output ...

	h_do_vector_add_destroy: 45 
	h_do_vector_add_destroy: 36 
	h_do_dot: 4785 
	h_do_dot: 3804 

*/

int
main( int ac, char *av[] )
{
	int i, k, *dp ;
	float *dbp ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;

	if (( k = cudaMalloc( &dp1, sizeof ( int ) * BUF_SIZE_INT )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	if (( k = cudaMalloc( &fp1, sizeof ( float ) * BUF_SIZE_INT )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	// h_expand_c_to_i testing 

	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		cbuf1[k] = 0x7b + k ;

	cudaMemcpy( dp1, cbuf1, BUF_SIZE_INT, cudaMemcpyHostToDevice ) ;

	h_cast_T_to_T<char, int> (( char *)dp1, ( int * )fp1, BUF_SIZE_INT ) ;
	dbg_p_d_data_i ("h_cast_T_to_T : char ",( int * )fp1, BUF_SIZE_INT );

	h_cast_T_to_T<unsigned char, int>(( unsigned char *)dp1, ( int * )fp1, BUF_SIZE_INT ) ;
	dbg_p_d_data_i ("h_cast_T_to_T : unsigned char ",( int * )fp1, BUF_SIZE_INT );

	h_cast_T_to_T<unsigned char, float>(( unsigned char *)dp1, ( float * )fp1, BUF_SIZE_INT ) ;
	dbg_p_d_data_f ("h_cast_T_to_T : to float",( float * )fp1, BUF_SIZE_INT );

	// test h_do_int_to_float : to diff buf

	dp = buf1 ;

	i = 0 ;
	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		*dp++ = i++ ;

	put_d_data_i( dp1, buf1, sizeof( int ) * BUF_SIZE_INT ) ;

	h_do_int_to_float ( dp1, fp1, BUF_SIZE_INT ) ;

	cudaMemcpy( dbuf, fp1, sizeof( float ) * BUF_SIZE_INT, cudaMemcpyDeviceToHost ) ;

	dbp = dbuf ;
	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		printf("diff buf -- %d : %f \n", k, *dbp++ ) ;

	// test h_do_int_to_float : to same buf

	memset( dbuf, 0, BUF_SIZE_INT * sizeof ( float )) ; // since int and float are of the same size

	dbp = dbuf ;
	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		printf("same buf -- %d : %f \n", k, *dbp++ ) ;

	dp = buf1 ;

	i = 4 ;
	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		*dp++ = i++ ;

	put_d_data_i( dp1, buf1, sizeof( int ) * BUF_SIZE_INT ) ;

	h_do_int_to_float ( dp1, (float *)dp1, BUF_SIZE_INT ) ;

	cudaMemcpy( dbuf, dp1, sizeof( float ) * BUF_SIZE_INT, cudaMemcpyDeviceToHost ) ;

	dbp = dbuf ;
	for ( k = 0 ; k < BUF_SIZE_INT ; k++ )
		printf("same buf -- %d : %f \n", k, *dbp++ ) ;
}
