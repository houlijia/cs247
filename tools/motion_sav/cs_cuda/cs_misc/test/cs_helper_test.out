Tue Mar  3 10:36:17 EST 2015
cs_helper_test
---------------------------------------------------
dbg_p_d_data_i: h_cast_T_to_T : char  size 40 dp 0x1300200200
dbg_pdata_i: h_cast_T_to_T : char 
0 -- 0000007b 123
1 -- 0000007c 124
2 -- 0000007d 125
3 -- 0000007e 126
4 -- 0000007f 127
5 -- ffffff80 -128
6 -- ffffff81 -127
7 -- ffffff82 -126
8 -- ffffff83 -125
9 -- ffffff84 -124
dbg_p_d_data_i: h_cast_T_to_T : unsigned char  size 40 dp 0x1300200200
dbg_pdata_i: h_cast_T_to_T : unsigned char 
0 -- 0000007b 123
1 -- 0000007c 124
2 -- 0000007d 125
3 -- 0000007e 126
4 -- 0000007f 127
5 -- 00000080 128
6 -- 00000081 129
7 -- 00000082 130
8 -- 00000083 131
9 -- 00000084 132
dbg_p_d_data_f: h_cast_T_to_T : to float size 40 dp 0x1300200200
dbg_pdata_f: h_cast_T_to_T : to float
0 -- 123.000000
1 -- 124.000000
2 -- 125.000000
3 -- 126.000000
4 -- 127.000000
5 -- 128.000000
6 -- 129.000000
7 -- 130.000000
8 -- 131.000000
9 -- 132.000000
diff buf -- 0 : 0.000000 
diff buf -- 1 : 1.000000 
diff buf -- 2 : 2.000000 
diff buf -- 3 : 3.000000 
diff buf -- 4 : 4.000000 
diff buf -- 5 : 5.000000 
diff buf -- 6 : 6.000000 
diff buf -- 7 : 7.000000 
diff buf -- 8 : 8.000000 
diff buf -- 9 : 9.000000 
same buf -- 0 : 0.000000 
same buf -- 1 : 0.000000 
same buf -- 2 : 0.000000 
same buf -- 3 : 0.000000 
same buf -- 4 : 0.000000 
same buf -- 5 : 0.000000 
same buf -- 6 : 0.000000 
same buf -- 7 : 0.000000 
same buf -- 8 : 0.000000 
same buf -- 9 : 0.000000 
same buf -- 0 : 4.000000 
same buf -- 1 : 5.000000 
same buf -- 2 : 6.000000 
same buf -- 3 : 7.000000 
same buf -- 4 : 8.000000 
same buf -- 5 : 9.000000 
same buf -- 6 : 10.000000 
same buf -- 7 : 11.000000 
same buf -- 8 : 12.000000 
same buf -- 9 : 13.000000 
